sdk/src/wallet/account/operations/balance.rs

```diff
@@ -4,178 +4,141 @@
     },
 };
 
 impl Account {
-    /// Get the AccountBalance
-    pub async fn balance(&self) -> crate::wallet::Result<AccountBalance> {
-        log::debug!("[BALANCE] get balance");
-        let mut account_balance = AccountBalance::default();
-        #[cfg(feature = "participation")]
-        {
-            account_balance.base_coin.voting_power = self.get_voting_power().await?;
-        }
+    /// Get the balance of the account.
+    pub async fn balance(&self) -> crate::wallet::Result<Balance> {
+        log::debug!("[BALANCE] balance");
 
-        let unlockable_outputs_with_multiple_unlock_conditions = self
-            .get_unlockable_outputs_with_additional_unlock_conditions(OutputsToClaim::All)
-            .await?;
-
-        let account_addresses = self.addresses().await?;
-
-        let network_id = self.client().get_network_id().await?;
-        let rent_structure = self.client().get_rent_structure().await?;
+        let account_details = self.details().await;
 
-        let local_time = self.client().get_time_checked().await?;
+        self.balance_inner(account_details.addresses_with_unspent_outputs.iter(), &account_details)
+            .await
+    }
 
-        let mut total_rent_amount = 0;
-        let mut total_native_tokens = NativeTokensBuilder::new();
+    /// Get the balance of the given addresses.
+    pub async fn addresses_balance(&self, addresses: Vec<impl Bech32AddressLike>) -> Result<Balance> {
+        log::debug!("[BALANCE] addresses_balance");
 
         let account_details = self.details().await;
 
-        let relevant_unspent_outputs = account_details
-            .unspent_outputs
-            .values()
-            // Check if output is from the network we're currently connected to
-            .filter(|data| data.network_id == network_id)
-            .map(|data| (&data.output_id, &data.output));
-
-        for (output_id, output) in relevant_unspent_outputs {
-            let rent = output.rent_cost(&rent_structure);
-
-            // Add alias and foundry outputs here because they can't have a [`StorageDepositReturnUnlockCondition`]
-            // or time related unlock conditions
-            match output {
-                Output::Alias(output) => {
-                    // Add amount
-                    account_balance.base_coin.total += output.amount();
-                    // Add storage deposit
-                    account_balance.required_storage_deposit.alias += rent;
-                    if !account_details.locked_outputs.contains(output_id) {
-                        total_rent_amount += rent;
-                    }
-
-                    // Add native tokens
-                    total_native_tokens.add_native_tokens(output.native_tokens().clone())?;
+        let addresses_with_unspent_outputs = addresses
+            .into_iter()
+            .map(|address| {
+                address
+                    .to_bech32()
+                    .map_err(|e| Error::Block(Box::new(e)))
```

Comment:
Can't refactor this by wrapping in `Ok`?

Reply:
How? The issue is the conversion

Reply:
Yeah but `and_then` require the same error type and there is no automatic conversion here

Reply:
You can't in the closure?

Reply:
Oh yeah, I guess it's a relic from another try, I remember trying ? and there was an issue, let me see